File: DotnetProject/SecurityDotnet/source/userAdmin/UserAdmin.cs
namespace SecurityDotnet.source.userAdmin;

public interface IUserAdmin
{
    UsersDTO? Login(string email, string password);
    List<TravelDTO>? GetTravels(string userId, string keyword);
    string RegisterUser(string email, string name, string password);
};

public class AspnetUsersDto
{
    public List<UsersDTO>? aspnetusers { get; set; }
}

public class UserAdmin : IUserAdmin
{
    private IMsIdentity _MsIdentity;

    public UserAdmin(IMsIdentity MsIdentity)
    {
        _MsIdentity = MsIdentity;
    }

    public UsersDTO? Login(string email, string password)
    {
        AspnetUsersDto? usersobj = JsonConvert.DeserializeObject<AspnetUsersDto>(File.ReadAllText("./database/aspnetUsers.json"));
        UsersDTO? user = usersobj?.aspnetusers?.FirstOrDefault(x => x.UserName == email);

        if (user == null)
            return null;

        if (_MsIdentity.ValidatePassword(password, user.PasswordHash))
            return user;

        return null;
    }

    public List<TravelDTO>? GetTravels(string userId, string keyword)
    {
        List<TravelDTO>? travels = JsonConvert.DeserializeObject<List<TravelDTO>>(File.ReadAllText("./database/travels.json"));
        List<TravelDTO>? result = travels?
            .FindAll(t => t.UserId == userId && t.From.Name.ToUpper().Contains(keyword.ToUpper()))
            .ToList();
        return result;
    }

    public string RegisterUser(string email, string name, string password)
    {

        var passwordHash = _MsIdentity.CreatePasswordHash(password);
        var newUser = new UsersDTO
        {
            Id = Guid.NewGuid().ToString(),
            FullName = name,
            UserName = email,
            PasswordHash = passwordHash
        };

        AspnetUsersDto? usersobj = JsonConvert.DeserializeObject<AspnetUsersDto>(File.ReadAllText("./database/aspnetUsers.json"));
        usersobj?.aspnetusers?.Add(newUser);

        File.WriteAllText(@"./database/aspnetUsers2.json", JsonConvert.SerializeObject(usersobj));

        return "Ok";
    }
}

File: NodeJsProject/msIdentity/useradmin.js
// Modul for å validere passord
const bcrypt = require('bcrypt');
const { readFileSync, writeFileSync } = require('fs');
const { newUserHashPassword, validatePassword } = require('./msIdentity');

// TODO Correct this filename after checking out
const databaseFolder = '../data/';
const databaseFile = 'aspnetUsers.json';
const travelsFile = 'travels.json'

exports.loadDatabase = loadDatabase = () => {
    const data = readFileSync(databaseFolder + databaseFile);
    jsonDB = JSON.parse(data)
    return jsonDB
}

exports.saveDatabase = saveDatabase = async (dbAsJson) => {
    let dbAsString = JSON.stringify(dbAsJson, null, '\t')
    const tt = writeFileSync(databaseFolder + databaseFile, dbAsString)
}

exports.getTravels = (userId, keyword) => {
    const jsonData = readFileSync(databaseFolder + travelsFile);
    data = JSON.parse(jsonData)
    let travels = data.filter(a => 
               a.UserId == userId &&
               (a.From.name.toUpperCase().includes(keyword.toUpperCase()) ||
               (a.To.name.toUpperCase().includes(keyword.toUpperCase()))))
    return travels
}

exports.getUser = getUser = (email) => {
    let db = this.loadDatabase()

    const user = db.aspnetusers.find(user => user.UserName === email);
    if (user) console.log(user)
    return user
}

exports.validateUser = validateUser = (email, password) => {
    try {        
        let user = getUser(email)        
        if (!user) 
            return false        
        let result = validatePassword(password, user.PasswordHash)                
        return result
    } catch (error) {
        console.log(error)
        return error
    } 
}

exports.registerUser = registerUser = (userName, fullName, passWord) => {
    // TODO hva om brukeren finnes
    try {
        let db = this.loadDatabase()
        let passwordHash = newUserHashPassword(passWord)
        
        let newUser = {
            "UserName": userName,
            "FullName": fullName,
            "PasswordHash": passwordHash
        }
        db.aspnetusers.push (newUser)
        console.log(db.aspnetusers)
        saveDatabase(db)
        return 'OK'
    } catch (error) {
        return error
    }
}